// @sap/approuter/lib/middleware/error-handler.js
'use strict';

const http = require('http');
const util = require('util');
const headerUtil = require('../utils/header-util');
const applicationLogUtils = require('../utils/application-logs-utils');
const dynamicRoutingUtils = require('../utils/dynamic-routing-utils');
const send = require('send');
const VError = require('verror').VError;
const setCircuitBreaker = require('./circuit-breaker-middleware').setCircuitBreaker;

const CONNECTION_ERROR_CODES = {
  ECONNRESET: true,
  ECONNREFUSED: true
};

module.exports = function errorHandler(err, req, res, next) { // eslint-disable-line no-unused-vars
  if (typeof err === 'string') {
    err = new Error(err);
  }
  const status = err.status || 500;
  setCircuitBreaker(req, err, status);
  const headers = err.headers || {};
  const tracer = req.loggingContext.getTracer(__filename);
  const logger = req.loggingContext.getLogger('/Handler');

  const errorMessage = req.method + ' request to ' + req.url + ' completed with status ' + status + ' ' + err.message;
  logger.error(errorMessage);
  applicationLogUtils.logRequestError(req, errorMessage);

  tracer.debug((err.stack ? ' stack: ' + err.stack : err));
  if (CONNECTION_ERROR_CODES[findErrCode(err)]) {
    tracer.info('Check your proxy and network settings.');
  }

  const routerConfig = req.routerConfig || req.app.get('mainRouterConfig');
  const redirectPage = routerConfig.appConfig.errorPage.get(status);
  if (redirectPage) {
    tracer.debug(`Redirecting to error page: ${redirectPage.file || redirectPage.path}`);
    if (redirectPage.file) {
      send(req, redirectPage.file, {index: false, root: routerConfig.workingDir})
        .on('error', function (err) {
          const verr = new VError(err, 'Unable to fetch custom error page: %s', redirectPage.file);
          endRequest(res, status, headers, verr);
        })
        .on('directory', function () {
          const verr = new VError(err, 'Custom error page is a directory: %s', redirectPage.file);
          endRequest(res, status, headers, verr);
        })
        .on('headers', function (res) {
          res.statusCode = status;
        })
        .pipe(res);
      return;
    } else if (redirectPage.path){
      if (!redirectPage.path.startsWith('/')) {
        redirectPage.path = '/' + redirectPage.path;
      }
      const url = dynamicRoutingUtils.getFullUrl(req, redirectPage.path);
      // Prevent endless loop in case error page itself returns an error
      let normalizedUrl = req.url;
      if (normalizedUrl.startsWith('//')) {
        normalizedUrl = normalizedUrl.substring(1);
      }
      if (normalizedUrl !== url) {
        res.writeHead(302, {
          Location: url
        });
        return res.end();
      }
    }
  }
  endRequest(res, status, headers, err);
};

function endRequest(res, status, headers, err) {
  if (res.headersSent) {
    return;
  }
  res.setHeader('Cache-Control', headerUtil.NOCACHE_HEADER_VALUE);
  res.writeHead(status, headers);
  let message = http.STATUS_CODES[status];
  if (err && process.env.NODE_ENV === 'development') {
    message = util.format('%d: %s\t%s', status, message, err.stack);
  }
  res.end(message);
}

function findErrCode(err) {
  if (!err) {
    return;
  }
  if ('code' in err) {
    return err.code;
  }
  if (typeof err.cause === 'function') {
    return findErrCode(err.cause());
  }
}
